sync-core/src/main/java/com/cloudant/sync/query/IndexUpdater.java

```diff
@@ -232,69 +233,68 @@ public Boolean call() {
+
+        if (arrayCount == 1) {
+            // Value is an array
             List<Object> arrayFieldValues;
-            arrayFieldValues = (ArrayList) ValueExtractor.extractValueForFieldName(arrayFieldName,
-                                                                                   rev.getBody());
-            for (Object value: arrayFieldValues) {
-                List<String> initialIncludedFields = new ArrayList<String>();
-                initialIncludedFields.add("_id");
-                initialIncludedFields.add("_rev");
-                initialIncludedFields.add(arrayFieldName);
-                List<Object> initialArgs = new ArrayList<Object>();
-                initialArgs.add(rev.getId());
-                initialArgs.add(rev.getRevision());
-                initialArgs.add(value);
-                DBParameter parameter;
-                parameter = populateDBParameter(fieldNames,
-                                                initialIncludedFields,
-                                                initialArgs,
-                                                indexName,
-                                                rev);
-                if (parameter == null) {
-                    return null;
+            arrayFieldValues = (List) ValueExtractor.extractValueForFieldName(arrayFieldName,
+                                                                              rev.getBody());
+            if (arrayFieldValues != null && arrayFieldValues.size() > 0) {
+                for (Object value: arrayFieldValues) {
+                    // For each value in the array we create a row. We put this value at the start
+                    // of the INSERT statement along with _id and _rev, followed by the other
+                    // fields. _id and _rev are special fields in that they don't appear in the
+                    // body, so they need special-casing to get the values.
+                    List<String> initialIncludedFields = Arrays.asList("_id",
+                                                                       "_rev",
+                                                                       arrayFieldName);
+                    List<Object> initialArgs = Arrays.asList(rev.getId(), rev.getRevision(), value);
+                    DBParameter parameter = populateDBParameter(fieldNames,
+                                                                initialIncludedFields,
+                                                                initialArgs,
+                                                                indexName,
+                                                                rev);
+                    if (parameter == null) {
+                        return null;
+                    }
+                    parameters.add(parameter);
                 }
-                parameters.add(parameter);
+                return  parameters;
```

Comment:
This early return statement belies confusing logic. The logic you're expressing is essentially an `if/else` block, with the if condition you want to alternate your logic being `if there was a non-empty array in the fields to be indexed`. However, the code doesn't make this obvious, instead resorting to an early return statement because of using the nested construct for what is logically a single condition.

Can you express this as a single if/else somehow, I think you need to do your checks in serial rather than nesting, like:

```
List<Object> arrayFieldValues = null;
if (arrayCount == 1) {
    arrayFieldValues = (List) ValueExtractor.extractValueForFieldName(arrayFieldName,
                                                                      rev.getBody());
}

if (arrayFieldValues != null && arrayFieldValues.size() > 0) {
    for (Object value: arrayFieldValues) {
      /* ... */
    }
    return parameters;
} else {
    List<String> initialIncludedFields = Arrays.asList("_id", "_rev");
    /* ... */
    return parameters;
}
```